pkg/promtail/targets/push/pushtarget.go

```diff
@@ -0,0 +1,203 @@
+	t.server.HTTP.Handle("/loki/api/v1/push", http.HandlerFunc(t.handle))
+
+	go func() {
+		err := srv.Run()
+		if err != nil {
+			level.Error(t.logger).Log("msg", "Loki push server shutdown with error", "err", err)
+		}
+	}()
+
+	return nil
+}
+
+func (t *PushTarget) handle(w http.ResponseWriter, r *http.Request) {
+	req, err := distributor.ParseRequest(r)
+	if err != nil {
+		level.Warn(t.logger).Log("msg", "failed to parse incoming push request", "err", err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	var lastErr error
+	for _, stream := range req.Streams {
+		matchers, err := logql.ParseMatchers(stream.Labels)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+
+		lb := labels.NewBuilder(make(labels.Labels, 0, len(matchers)+len(t.config.Labels)))
+
+		// Add configured labels
+		for k, v := range t.config.Labels {
+			lb.Set(string(k), string(v))
+		}
+
+		// Add stream labels
+		for i := range matchers {
+			lb.Set(matchers[i].Name, matchers[i].Value)
+		}
+
+		// Apply relabeling
+		processed := relabel.Process(lb.Labels(), t.relabelConfig...)
+		if processed == nil || len(processed) == 0 {
+			w.WriteHeader(http.StatusNoContent)
+			return
+		}
+
+		// Convert to model.LabelSet
+		filtered := model.LabelSet{}
+		for i := range processed {
+			if strings.HasPrefix(processed[i].Name, "__") {
+				continue
+			}
+			filtered[model.LabelName(processed[i].Name)] = model.LabelValue(processed[i].Value)
+		}
+
+		for _, entry := range stream.Entries {
+			var err error
+			if t.config.KeepTimestamp {
+				err = t.handler.Handle(filtered, entry.Timestamp, entry.Line)
+			} else {
+				err = t.handler.Handle(filtered, time.Now(), entry.Line)
+			}
+
+			if err != nil {
+				lastErr = err
+				continue
+			}
+		}
+	}
+
+	if lastErr != nil {
+		level.Warn(t.logger).Log("msg", "at least one entry in the push request failed to process", "err", lastErr.Error())
+		http.Error(w, lastErr.Error(), http.StatusBadRequest)
+		return
+	}
+
+	w.WriteHeader(http.StatusNoContent)
+	return
+}
+
+// Type returns PushTargetType.
+func (t *PushTarget) Type() target.TargetType {
+	return target.PushTargetType
+}
+
+// Ready indicates whether or not the PushTarget target is ready to be read from.
+func (t *PushTarget) Ready() bool {
+	return true
+}
+
+// DiscoveredLabels returns the set of labels discovered by the PushTarget, which
+// is always nil. Implements Target.
+func (t *PushTarget) DiscoveredLabels() model.LabelSet {
+	return nil
+}
+
+// Labels returns the set of labels that statically apply to all log entries
+// produced by the PushTarget.
+func (t *PushTarget) Labels() model.LabelSet {
+	return t.config.Labels
+}
+
+// Details returns target-specific details.
+func (t *PushTarget) Details() interface{} {
+	return map[string]string{}
+}
+
+// Stop shuts down the PushTarget.
+func (t *PushTarget) Stop() error {
+	level.Info(t.logger).Log("msg", "stopping push server", "job", t.jobName)
+	t.server.Shutdown()
```

Comment:
Wanted to add a note, not sure here if it's worth having a waitgroup on the server thread and wait for it before returning here?